### ISSUE TITLE
dart2js crash compiling language/await/future_test

### ISSUE THREAD

**Issue by munificent** [dart-lang/sdk#51677]
With records and patterns enabled, I see [a crash in dart2js](https://logs.chromium.org/logs/dart/buildbucket/cr-buildbucket/8787157786924699617/+/u/test_results/new_test_failures__logs_) when compiling this test:

```
/===========================================================================================\
| dart2js-minified-linux-d8:language/await/future_test broke (Pass -> Crash, expected Pass) |
\===========================================================================================/

--- Command "dart2js" (took 03.000843s):
DART_CONFIGURATION=ReleaseX64 out/ReleaseX64/dart-sdk/bin/dart compile js -Dtest_runner.configuration=dart2js-minified-linux-d8 --sound-null-safety --test-mode --minify --packages=/b/s/w/ir/.dart_tool/package_config.json /b/s/w/ir/tests/language/await/future_test.dart --sound-null-safety --out=/b/s/w/ir/out/ReleaseX64/generated_compilations/dart2js-minified-linux-d8/tests_language_await_future_test/future_test.js

exit code:
253

stdout:
Info: Compiling with sound null safety.
tests/language/await/future_test.dart:117:1:
Internal Error: The compiler crashed when compiling this element.
awaitSwitch() async {
^
The compiler is broken.

When compiling the above element, the compiler crashed. It is not
possible to tell if this is caused by a problem in your program or
not. Regardless, the compiler should not crash.

The Dart team would greatly appreciate if you would take a moment to
report this problem at http://dartbug.com/new.

Please include the following information:

* the name and version of your operating system,

* the Dart SDK build number (3.0.0-edge.18d414d24dfe537d3b366b2f4a08a3cb469aced1), and

* the entire message you see here (including the full stack trace
  below as well as the source location above).

The compiler crashed: Null check operator used on a null value
#0      HBasicBlock.assignCommonDominator (package:compiler/src/ssa/nodes.dart:1006:36)
#1      HGraph.assignDominators (package:compiler/src/ssa/nodes.dart:378:17)
#2      HGraph.finalize (package:compiler/src/ssa/nodes.dart:364:5)
#3      KernelSsaGraphBuilder._closeFunction (package:compiler/src/ssa/builder.dart:1817:11)
#4      KernelSsaGraphBuilder._buildGenerator (package:compiler/src/ssa/builder.dart:1414:9)
#5      KernelSsaGraphBuilder._buildFunctionNode (package:compiler/src/ssa/builder.dart:1323:7)
#6      KernelSsaGraphBuilder.build.<anonymous closure> (package:compiler/src/ssa/builder.dart:469:15)
#7      DiagnosticReporter.withCurrentElement (package:compiler/src/diagnostics/diagnostic_listener.dart:142:15)
#8      KernelSsaGraphBuilder.build (package:compiler/src/ssa/builder.dart:454:21)
#9      KernelSsaBuilder.build.<anonymous closure> (package:compiler/src/js_model/js_strategy.dart:545:22)
#10     CompilerTask.measure (package:compiler/src/common/tasks.dart:66:51)
#11     KernelSsaBuilder.build (package:compiler/src/js_model/js_strategy.dart:527:18)
#12     SsaBuilderTask.build (package:compiler/src/ssa/ssa.dart:361:21)
#13     SsaFunctionCompiler.compile (package:compiler/src/ssa/ssa.dart:87:28)
#14     OnDemandCodegenResults.getCodegenResults (package:compiler/src/common/codegen.dart:421:30)
#15     JsBackendStrategy.generateCode (package:compiler/src/js_model/js_strategy.dart:334:43)
#16     KernelCodegenWorkItem.run (package:compiler/src/js_model/js_strategy.dart:486:29)
#17     Compiler.emptyQueue.<anonymous closure>.<anonymous closure>.<anonymous closure>.<anonymous closure>.<anonymous closure> (package:compiler/src/compiler.dart:822:70)
#18     CompilerTask.measureSubtask (package:compiler/src/common/tasks.dart:181:35)
#19     Compiler.emptyQueue.<anonymous closure>.<anonymous closure>.<anonymous closure>.<anonymous closure> (package:compiler/src/compiler.dart:822:32)
#20     CompilerTask.measureSubtask (package:compiler/src/common/tasks.dart:181:35)
#21     Compiler.emptyQueue.<anonymous closure>.<anonymous closure>.<anonymous closure> (package:compiler/src/compiler.dart:820:28)
#22     DiagnosticReporter.withCurrentElement (package:compiler/src/diagnostics/diagnostic_listener.dart:142:15)
#23     Compiler.emptyQueue.<anonymous closure>.<anonymous closure> (package:compiler/src/compiler.dart:818:18)
#24     CodegenEnqueuer._forEach (package:compiler/src/js_backend/enqueuer.dart:252:12)
#25     CodegenEnqueuer.forEach (package:compiler/src/js_backend/enqueuer.dart:268:5)
#26     Compiler.emptyQueue.<anonymous closure> (package:compiler/src/compiler.dart:814:16)
#27     CompilerTask.measureSubtask (package:compiler/src/common/tasks.dart:181:35)
#28     Compiler.emptyQueue (package:compiler/src/compiler.dart:813:14)
#29     Compiler.processQueue.<anonymous closure> (package:compiler/src/compiler.dart:837:7)
#30     CompilerTask.measureSubtask (package:compiler/src/common/tasks.dart:181:35)
#31     Compiler.processQueue (package:compiler/src/compiler.dart:831:14)
#32     Compiler.runCodegenEnqueuer (package:compiler/src/compiler.dart:551:5)
#33     Compiler.runSequentialPhases (package:compiler/src/compiler.dart:763:23)
<asynchronous suspension>
#34     Compiler.runInternal.<anonymous closure> (package:compiler/src/compiler.dart:318:7)
<asynchronous suspension>
#35     Compiler.runInternal (package:compiler/src/compiler.dart:317:5)
<asynchronous suspension>
#36     Compiler.run.<anonymous closure> (package:compiler/src/compiler.dart:238:11)
<asynchronous suspension>

--- Re-run this test:
python3 tools/test.py -n dart2js-minified-linux-d8 language/await/future_test
```

cc @rakudrama

**Comment by vsmenon** [comment]
Marking this a P1 since it's a crasher.

**Comment by rakudrama** [comment]
Fixed by 71d6292eb30a7e8b19284231f01dd2dc560c283d
